helix-core/src/test/java/org/apache/helix/model/TestClusterConfig.java

```diff
@@ -122,13 +122,6 @@ public void testGetRebalancePreferenceDefault() {
     ClusterConfig testConfig = new ClusterConfig("testId");
     Assert.assertEquals(testConfig.getGlobalRebalancePreference(),
         ClusterConfig.DEFAULT_GLOBAL_REBALANCE_PREFERENCE);
-
-    Map<ClusterConfig.GlobalRebalancePreferenceKey, Integer> preference = new HashMap<>();
```

Comment:
What would be the expected preference in this condition then?

Reply:
Moreover, if we move the auto-fill logic to ConstraintBasedAlgorithmFactory.java, then we need to add a test case there to cover the change.

Reply:
They would be able to successfully set the preference, which is already tested by the above test case. 

I can add a test case in ConstraintBasedAlgorithmFactory for this.

Reply:
On a second thought, do you think it's actually better to move the validation logic in ConstraintBasedAlgorithmFactory? My reasoning for the current approach is that we shouldn't make a special case out of certain keys because I want to keep the solution generic, but doing the check in ConstraintBasedAlgorithmFactory is just delaying the same thing. 

It might be better to keep the validation in setter since it's not silent.

Reply:
Either way works. But we need to make it clean. The validation should be done in one place. So ConstraintBasedAlgorithmFactory might be a better option given the logic becomes more complex. But I don't have a strong preference.

Reply:
Moving back to ClusterConfig to validate during `set()`.